ops/pebble.py

```diff
@@ -660,6 +660,42 @@ def to_dict(self) -> typing.Dict[str, typing.Any]:
         ]
         return {name: value for name, value in fields if value}
 
+    def merge(self, service: 'Service'):
+        """Merges this service object with another service definition."""
+        for key in service.to_dict():
+            # Setup an anonymous function that will update the service definition.
+            if key == 'after':
+                setter = setattr(self, key, self.after + service.after)
+            elif key == 'before':
+                setter = setattr(self, key, self.before + service.before)
+            elif key == 'requires':
+                setter = setattr(self, key, self.requires + service.requires)
+            elif key == 'environment':
+                def setter():
+                    return None
+                for key in service.environment:
+                    self.environment[key] = service.environment[key]
+            elif key == 'on_success':
+                def setter():
+                    return None
+                for key in service.on_success:
+                    self.on_success[key] = service.on_success[key]
+            elif key == 'on_failure':
+                def setter():
+                    return None
+                for key in service.on_failure:
+                    self.on_failure[key] = service.on_failure[key]
+            elif key == 'on_check_failure':
+                def setter():
+                    return None
+                for key in service.on_check_failure:
+                    self.on_check_failure[key] = service.on_check_failure[key]
+            else:
+                def setter():
+                    return setattr(self, key, getattr(service, key))
+
+            setter()
+
```

Comment:
I think the current implementation has several bugs, presumably none of which were caught with the test you've added (presumably because it happens to dodge the problematic fields :-):

* The `to_dict` method returns the JSON/YAML key names, which differ from the Python attributes names. For example, `on_success` becomes `on-success`. So `setattr` won't work as expected for those.
* `setter = setattr(...)` actually calls the `setattr` function, but then you try to call the result of that further down -- presumably that second call will fail with a `'NoneType' is not callable` error.
* The `on-success` and `on-failure` fields are scalar, not maps (`on-check-failure` is a map, however).

Other notes:

* With the Go implementation we had to hand-roll it because Go is statically typed (we did consider reflection and code generation, but decided against them in the end). However, Python is very dynamic so I think we can do something simpler here -- see comment below for suggestions:
* If we sticking with this approach, I think it'd be clearer to just do the attribute setting in the `if` block -- the function indirection just obfuscates it. And you wouldn't need the odd `def setter(): return None` dance then either.
* Python's `dict.items()` is handy here and slightly simplifies the dict loops. Also consider `dict.update()`.
* Naming nit: I think "service" should be called "other", otherwise it's a bit confusing what service it is. "other" denotes it's not self, but the *other* one you're merging in.

Reply:
We can use `__dict__` in place of `to_dict()`. We can simplify and generalize by testing the type rather than the field name. So here's one approach that seems to work in my quick tests:

```python
class Service:
    ...

    def _merge(self, other: 'Service'):
        for name, value in other.__dict__.items():
            if not value:
                continue
            if isinstance(value, list):
                getattr(self, name).extend(value)
            elif isinstance(value, dict):
                getattr(self, name).update(value)
            else:
                setattr(self, name, value)
```

The above would actually work with any similar type, not just `Service`. Here's another approach that's not generalized, but uses the field names explicitly, grouping together similar fields:

```python
    def _merge(self, other: 'Service'):
        for name, value in other.__dict__.items():
            if not value:
                continue
            if name in ['after', 'before', 'requires']:
                getattr(self, name).extend(value)
            elif name in ['environment', 'on_check_failure']:
                getattr(self, name).update(value)
            else:
                setattr(self, name, value)
```

Reply:
Since we are looping over the "other" service, we probably need to guard against the attrib not existing in `self.__dict__` - e.g.:

```python
if name in ['after', 'before', 'requires'] and hasattr(self, name):
    getattr(self, name).extend(value)
elif name in ['environment', 'on_check_failure'] and hasattr(self, name):
    getattr(self, name).update(value)
...
```

Reply:
Does this actually matter? Because `other` is another `Service` object, so will have the same attributes, right?

Reply:
It shouldn't matter. Both `Service` objects should have all the properties. Those with null values are left out of `to_dict`, but they still exist in the class.

Reply:
Hah! - guilty of thinking _too_ dynamically :P